### DISCUSSION TITLE
Spectrograms

### DISCUSSION THREAD

**Discussion by ghost** [mcguirepr89/BirdNET-Pi#8] [Ideas]
`sox PathToWaveFile.wav -n spectrogram -o spec.png` generates spectrograms:

![spec](https://user-images.githubusercontent.com/78355888/136189281-f1b97f2b-74a1-481b-9b8f-e88b010b78df.png)

Perhaps this feature could be easyly included in the birdsystem.local browser.

**Comment by mcguirepr89** [comment]
I like this -- that feature is built into sox????

**Reply by ghost** [reply]
yes

**Reply by mcguirepr89** [reply]
I just tested that myself and you are right -- this ought to be incorporated somehow. Set me see what I can come up with

**Reply by mcguirepr89** [reply]
I just added that feature to the latest commit to rpialpha if you wanted to test?

**Reply by mcguirepr89** [reply]
ahh wait it needs a bit more -- give me a few more minutes

**Reply by mcguirepr89** [reply]
ok now it is ready

**Reply by ghost** [reply]
nice, works. I continue with the endurance testing.

**Reply by mcguirepr89** [reply]
Thank you, friend -- great suggestion and speedy testing

**Reply by DD4WH** [reply]
My argument for this is written up here:
https://github.com/mcguirepr89/BirdNET-Pi/issues/33#issuecomment-944920817

**Comment by mcguirepr89** [comment]
![Screenshot from 2021-10-16 19-01-32](https://user-images.githubusercontent.com/60325264/137604089-82971aa4-4add-4c96-b0a6-8148615bdded.png)


I was hoping that the spectrograms would help me with these nearly indiscernably different species:
![Screenshot from 2021-10-16 18-59-17](https://user-images.githubusercontent.com/60325264/137604059-559cbd92-a85d-43b7-ac91-5a8696b270bd.png)
. . . naivete

**Comment by krummrey** [comment]
I've been looking into the sox options trying to see what one can change. Here are my findings:

Old:
![spectrogram-alt](https://user-images.githubusercontent.com/4517632/185933774-56344812-c806-4346-9b31-95f9e68479f4.png)

New:
![spectrogram-neu-A](https://user-images.githubusercontent.com/4517632/185933781-5f4001e6-2275-42ab-832b-3c7ee5704c4e.png)

`sox -V1 birdnet.mp3 -n remix 1 rate 24k spectrogram -p 2 -w Dolph -Y 320 -z 75 -Z -32  -A -c birdenet.mp3 -o spectrogram-neu-A.png`
This is what I've changed: 

- made the height a little smaller
- changed the color scheme 
(fits more to the green of birdnetpi)
- changed the input range to suit my shabby microphone 
(this probybly needs to be adjusted for every microphone)

Not a huge difference, like [soundshader](https://soundshader.github.io/hss/gallery/) would be.

**Reply by matthew73210** [reply]
Ah this could be used as the basis to a new idea, why not let the user have the option to change and tune the spectograph? Colours etc... Even go from time domain to frequency which would add a lot more options for signal processing (I'm a geek so this sort of stuff, albeit useless for most, find it very interesting to do even if I get no useful information)
I know that this can be done trough the CL but some people don't know how to.

**Reply by DCS17** [reply]
I'm wondering how to change the color of the live spectrogram player viewer?
I've been using @krummrey suggestion regaring sox options and I much prefer the blue/green to the purple - great job!

I've managed to change the "currently analyzing" and extracted spectrograms by editing the spectrogram.sh file and extract_new_birdsongs.sh files. 
But nothing I have tried will cause the live spectrogram scolling player to change. I can't figure out where it is getting its scrolling image from?
I'm not a programmer, so really have no idea what I'm doing - if that's not already obvious :-)
But I'm happy to tinker until I break something :-)

Thanks for any advice.

**Reply by ehpersonal38** [reply]
@DCS17 HI there! I certainly think this is achievable with a little tinkering. The file to edit is scripts/spectrogram.php, and the lines to try changing are around 305. I think this bit of code is where I'd start: `let hue = Math.round((rat * 120) + 280 % 360);`

 I'd start by changing the number on the very end of that line (360), then working inwards if that doesn't produce the desired effect, if that makes sense. Let me know if you get it working!

**Reply by DCS17** [reply]
Thanks for your quick reply. I would never have worked that out for myself - so thank you! 
I've spent a bit of time playing and trying to learn !
Definitely made some changes that I prefer, but I can't quite get the same that the sox output gives. But near enough!
Plenty of tinkering left to try though :-)

**Reply by ehpersonal38** [reply]
You're welcome! :) 

The person who wrote the code behind the spectrogram actually uploaded a video showing how he coded the whole thing. Here's a link to the timestampped part where he explains the colors: https://youtu.be/hYNJGPnmwls?t=1365

I messed around a bit after watching that vid but admittedly I got a C in high school math so it's a little too far over my head. Could be helpful to you if you want to try tinkering more though? Best of luck!

(oh also, I think I was wrong to say you should focus on changing 360 first, I think 360 needs to always stay that value since it represents the total number of hues)

**Reply by DCS17** [reply]
Yes. I've been messing and found that the 120/280 was the best to change. I'll take a look at the video and brush up on my math! Thanks.

**Reply by DCS17** [reply]
The video link sorted it out in my head. Thanks.
I've now got 
CTX.fillStyle = 'hsl(240, 100%, 10%)';
with
let hue = Math.round((-rat * 220) + 240);
and
let lit = 10 + (85 * rat) + '%';

looking pretty much identical to the blue/green sox spectrogram above.
So very happy :-)
Thanks for your help.

**Reply by ehpersonal38** [reply]
![chrome_ypZ8R5BSN5](https://user-images.githubusercontent.com/103586016/189550609-0b06c76b-32d6-4212-836e-c5630b7f5d20.png)


That's super cool!! Glad we were able to arrive at a solution! Also FYI you will probably need to adjust line 137 of the spectrogram code to update the fillStyle there too (change the 280 to 240).

**Reply by DCS17** [reply]
Not sure what line 137 does. I hadn't changed it (hadn't noticed it!) and all seemed OK. Anyway I've change it now too. All good 👍

**Comment by JP-FKA** [comment]
Now that this nice feature is implemented in the web pages, I wonder what are the Gain and Compression options.
Gain allows to diminish volume, but I don't know if it is set on the hardware acquisition, or later in the chain.
Compression: i don't understand what that means

**Reply by matthew73210** [reply]
> Now that this nice feature is implemented in the web pages, I wonder what are the Gain and Compression options. Gain allows to diminish volume, but I don't know if it is set on the hardware acquisition, or later in the chain. Compression: i don't understand what that means

If I'm not mistaken, it changes what sort of compression is used. Which can affect the recording in multiple ways. Think of .mp3 files.

**Reply by ehpersonal38** [reply]
The live spectrogram is completely separate from your recordings, so changing the gain will only change the volume of the live stream in-browser. Compression is just a filter that tries to level quiet sounds and loud sounds to be more equal in volume.
